src/CoffeeShop.java

```diff
@@ -0,0 +1,85 @@
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+
+public class CoffeeShop {
+    String name;
+    double x;
+    double y;
+    double distance;
+
+    public CoffeeShop(String name, double x, double y, double distance) {
+        this.name = name;
+        this.x = x;
+        this.y = y;
+        this.distance = distance;
+    }
+
+    // main method, expecting 3 command-line arguments (user x coords, user y coords, URL of the coffee shop CSV file)
+    public static void main(String[] args) {
+        if(args.length != 3){
+            System.out.println("Usage: <user_x_coordinate> <user_y_coordinate> <shop_data_url>");
+            return;
+        }
+        try{
+            // parse user coordinates
+            double x = Double.parseDouble(args[0]);
+            double y = Double.parseDouble(args[1]);
+            String shopUrl = args[2];
+
+            //fetch coffee shops and calculates the distance
+            List<CoffeeShop> coffeeShopList = fetchCoffeeShops(shopUrl, x, y);
+
+            // sort shops by distance
+            coffeeShopList.sort(Comparator.comparingDouble(shop -> shop.distance));
+
+            int count = Math.min(3, coffeeShopList.size()); // print top 3 closest coffee shops
+            for(int i = 0; i < count; i++){
+                CoffeeShop shop = coffeeShopList.get(i);
+                System.out.printf("%s,%.4f\n",shop.name, shop.distance);
+            }
+        } catch(Exception e){
+            System.out.println("Error: " + e.getMessage());
+            e.printStackTrace();
+        }
+    }
+
+    public static List<CoffeeShop> fetchCoffeeShops(String url, double userX, double userY) throws Exception{
+        List<CoffeeShop> shops = new ArrayList<>(); // list for storing coffee shops
+        URL url1 = new URL(url);
+        BufferedReader reader = new BufferedReader(new InputStreamReader(url1.openStream())); // using buffer reader to read from urls
+        String line;
+
+        while((line = reader.readLine()) != null){
+            String[] parts = line.split(",");
+
+            // skip invalid lines that don't match the expected format
+            if(parts.length != 3){
```

Comment:
👍🏽 for skipping invalid data and not abandoning the entire process for invalid records.